flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/NativeFlinkService.java

```diff
@@ -306,34 +310,42 @@ protected Map<JobVertexID, JobVertexResourceRequirements> getVertexResources(
     }
 
     /**
-     * Scale JM deployment to zero to gracefully stop all JM instances before any TMs are stopped.
-     * This avoids race conditions between JM shutdown and TM shutdown / failure handling.
+     * Shut down JobManagers gracefully by scaling JM deployment to zero. This avoids race
+     * conditions between JM shutdown and TM shutdown / failure handling.
      *
      * @param jmDeployment
      * @param namespace
      * @param clusterId
-     * @param timeout
+     * @param remainingTimeout
      * @return Remaining timeout after the operation.
      */
-    private Duration scaleJmToZeroBlocking(
+    private Duration shutdownJobManagers(
             EditReplacePatchable<Deployment> jmDeployment,
             String namespace,
             String clusterId,
-            Duration timeout) {
-        return deleteBlocking(
-                "Scaling JobManager Deployment to zero",
-                () -> {
-                    try {
-                        jmDeployment.patch(PatchContext.of(PatchType.JSON_MERGE), SCALE_TO_ZERO);
-                    } catch (Exception ignore) {
-                        // Ignore all errors here as this is an optional step
-                        return null;
-                    }
-                    return kubernetesClient
-                            .pods()
-                            .inNamespace(namespace)
-                            .withLabels(KubernetesUtils.getJobManagerSelectors(clusterId));
-                },
-                timeout);
+            Duration remainingTimeout) {
+
+        // We use only half of the shutdown timeout but at most one minute as the main point
+        // here is to initiate JM shutdown before the TMs
+        var jmShutdownTimeout =
+                ObjectUtils.min(JM_SHUTDOWN_MAX_WAIT, remainingTimeout.dividedBy(2));
```

Comment:
I think simply issuing a non-blocking delete call after the entire wait period would suffice.

Reply:
I don't fully understand what you mean by non-blocking delete. What we gain here by shutting down the JM first is cleaner logs and generally faster shutdown for the TM pods which overall leads to more stable shutdown for the entire job.

Reply:
This is not really user facing logic in the sense that the only thing that user cares about is a clean shutdown within the configured total shutdown timeout.

Reply:
The reason why clean blocking deletion is always preferred is due to issues like https://issues.apache.org/jira/browse/FLINK-32334 .
It is always better to wait for the deployments/pods to go away before doing anything otherwise strange things can happen, like JMs/TMs reconnecting to new clusters etc

Reply:
I think I understand what the new behavior is trying to solve. I'm just saying that there is no need to time out the scale down request after half the timeout (or after one minute). The old behavior (in the main branch) is sufficient and avoids unnecessary complexity.

Reply:
I'm proposing to keep the core deletion logic as in the main branch, together with the other changes in this PR (condition on Foreground deletion).

Reply:
I was missing that a Foreground delete call would clean up TaskManagers through delete propagation, even if the JM pod was stuck.